File: python/equistore-torch/tests/block.py
from typing import Dict, List

import pytest
import torch
from torch import Tensor

from equistore.torch import Labels, TensorBlock


def test_constructor():
    # keyword arguments
    block = TensorBlock(
        values=torch.full((3, 2), 11),
        samples=Labels(names=["s"], values=torch.IntTensor([[0], [2], [1]])),
        components=[],
        properties=Labels(names=["p"], values=torch.IntTensor([[1], [0]])),
    )

    assert torch.all(block.values == torch.full((3, 2), 11))
    assert block.samples.names == ["s"]
    assert block.components == []
    assert block.properties.names == ["p"]

    # positional arguments
    block = TensorBlock(
        torch.full((3, 2), 33),
        Labels(names=["s"], values=torch.IntTensor([[0], [2], [1]])),
        [],
        Labels(names=["p"], values=torch.IntTensor([[1], [0]])),
    )

    assert torch.all(block.values == torch.full((3, 2), 33))
    assert block.samples.names == ["s"]
    assert block.components == []
    assert block.properties.names == ["p"]


def test_clone():
    values = torch.full((3, 2), 11)
    block = TensorBlock(
        values=values,
        samples=Labels(names=["s"], values=torch.IntTensor([[0], [2], [1]])),
        components=[],
        properties=Labels(names=["p"], values=torch.IntTensor([[1], [0]])),
    )

    assert values.data_ptr() == block.values.data_ptr()

    clone = block.copy()
    del block

    assert values.data_ptr() != clone.values.data_ptr()


def test_gradients():
    block = TensorBlock(
        values=torch.full((3, 2), 1),
        samples=Labels(names=["s"], values=torch.IntTensor([[0], [2], [1]])),
        components=[],
        properties=Labels(names=["p"], values=torch.IntTensor([[1], [0]])),
    )

    assert block.gradients_list() == []

    block.add_gradient(
        parameter="g",
        gradient=TensorBlock(
            values=torch.full((1, 3, 2), 11),
            samples=Labels(
                names=("sample", "g"),
                values=torch.IntTensor([[0, 1]]),
            ),
            components=[Labels(names=("c"), values=torch.IntTensor([[0], [1], [2]]))],
            properties=block.properties,
        ),
    )

    assert block.gradients_list() == ["g"]
    assert block.has_gradient("g") is True
    assert block.has_gradient(parameter="not-there") is False

    gradient = block.gradient("g")
    assert gradient.values.shape == (1, 3, 2)

    gradient = block.gradient(parameter="g")
    assert gradient.samples.names == ["sample", "g"]

    message = "can not find gradients with respect to 'not-there' in this block"
    with pytest.raises(RuntimeError, match=message):
        gradient = block.gradient("not-there")

    gradients = block.gradients()
    assert isinstance(gradients, dict)
    assert list(gradients.keys()) == ["g"]


# define a wrapper class to make sure the types TorchScript uses for of all
# C-defined functions matches what we expect
class TensorBlockWrap:
    def __init__(
        self,
        values: Tensor,
        samples: Labels,
        components: List[Labels],
        properties: Labels,
    ):
        self._c = TensorBlock(
            values=values,
            samples=samples,
            components=components,
            properties=properties,
        )

    def copy(self) -> TensorBlock:
        return self._c.copy()

    def values(self) -> Tensor:
        return self._c.values

    def samples(self) -> Labels:
        return self._c.samples

    def components(self) -> List[Labels]:
        return self._c.components

    def properties(self) -> Labels:
        return self._c.properties

    def add_gradient(self, parameter: str, gradient: TensorBlock):
        return self._c.add_gradient(parameter=parameter, gradient=gradient)

    def gradients_list(self) -> List[str]:
        return self._c.gradients_list()

    def has_gradient(self, parameter: str) -> bool:
        return self._c.has_gradient(parameter=parameter)

    def gradient(self, parameter: str) -> TensorBlock:
        return self._c.gradient(parameter=parameter)

    def gradients(self) -> Dict[str, TensorBlock]:
        return self._c.gradients()


def test_script():
    class TestModule(torch.nn.Module):
        def forward(self, x: TensorBlockWrap) -> TensorBlockWrap:
            return x

    module = TestModule()
    module = torch.jit.script(module)

File: equistore-torch/tests/block.cpp
#include <torch/torch.h>

#include <equistore/torch.hpp>
using namespace equistore_torch;

#include <catch.hpp>

TEST_CASE("Blocks") {
    SECTION("constructors and data") {
        auto block = TensorBlockHolder(
            torch::full({3, 2}, 11.0),
            LabelsHolder::create({"s"}, {{0}, {2}, {1}}),
            {},
            LabelsHolder::create({"p"}, {{0}, {1}})
        );

        CHECK((block.values().sizes() == std::vector<int64_t>{3, 2}));
        CHECK(torch::all(block.values() == 11.0).item<bool>());

        CHECK(*block.samples() == equistore::Labels({"s"}, {{0}, {2}, {1}}));
        CHECK(block.samples()->names().toTupleRef().size() == 1);
        CHECK(block.samples()->names().toTupleRef().elements()[0].toStringRef() == "s");

        CHECK(block.components().empty());

        CHECK(*block.properties() == equistore::Labels({"p"}, {{0}, {1}}));
        CHECK(block.properties()->names().toTupleRef().size() == 1);
        CHECK(block.properties()->names().toTupleRef().elements()[0].toStringRef() == "p");
    }

    SECTION("clone") {
        auto values = torch::full({3, 2}, 11.0);
        auto block = TensorBlockHolder(
            values,
            LabelsHolder::create({"s"}, {{0}, {2}, {1}}),
            {},
            LabelsHolder::create({"p"}, {{0}, {1}})
        );

        CHECK(values.data_ptr() == block.values().data_ptr());

        auto clone = block.copy();
        CHECK(values.data_ptr() != clone->values().data_ptr());
        CHECK(torch::all(values == clone->values()).item<bool>());
    }

    SECTION("gradients") {
        auto block = TensorBlockHolder(
            torch::full({3, 2}, 11.0),
            LabelsHolder::create({"s"}, {{0}, {2}, {1}}),
            {},
            LabelsHolder::create({"p"}, {{0}, {1}})
        );

        CHECK(block.gradients_list().empty());

        block.add_gradient("g", torch::make_intrusive<TensorBlockHolder>(
            torch::full({1, 3, 2}, 1.0),
            LabelsHolder::create({"sample", "g"}, {{0, 1}}),
            std::vector<TorchLabels>{LabelsHolder::create({"c"}, {{0}, {1}, {2}})},
            block.properties()
        ));

        CHECK((block.gradients_list() == std::vector<std::string>{"g"}));
        CHECK(block.has_gradient("g"));
        CHECK_FALSE(block.has_gradient("not-there"));

        auto gradient = block.gradient("g");
        CHECK((gradient->values().sizes() == std::vector<int64_t>{1, 3, 2}));

        auto sample_names = gradient->samples()->names().toTupleRef().elements();
        CHECK(sample_names.size() == 2);
        CHECK(sample_names[0].toStringRef() == "sample");
        CHECK(sample_names[1].toStringRef() == "g");

        auto gradients = block.gradients().toGenericDict();

        for (const auto& entry: gradients) {
            CHECK(entry.key().toStringRef() == "g");
        }
    }
}
